### PR TITLE
PBC parsing and comparing utility

### PR DESCRIPTION
**Pull request by pszczesniak**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal (utils): Frontend licence parsing and comparing utility.
Closes https://github.com/cksource/ckeditor5-internal/issues/3164.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-utils/tests/madewith.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+			} );
+		} );
+
+		describe( 'should return `INVALID`', () => {
+			it( 'when passed variable is invalid', () => {
+				// invalid
+				const string = 'foobarbaz';
+
+				expect( verify( string ) ).to.be.equal( 'INVALID' );
+			} );
```
Comment (Reinmar) [inline packages/ckeditor5-utils/tests/madewith.js:32]:
```suggestion
			} );
```

File: packages/ckeditor5-utils/src/madewith.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+export type VerifiedKeyStatus = 'VALID' | 'INVALID' | 'EXPIRED';
+
+/**
+ * Checks whether the given string contains information that allows you to verify the license status.
+ *
+ * @param stringToCheck The string to check.
+ * @returns String that represents the state of given `stringToCheck` parameter. It can be `'VALID'`, `'INVALID'` or `'EXPIRED'`.
+ */
+export default function verify( stringToCheck: string ): VerifiedKeyStatus {
+	// This is just a very simplified preliminary front-end check of the date validation with the current release date - it
```
Comment (Reinmar) [inline packages/ckeditor5-utils/src/madewith.ts:22]: What's the purpose of this comment?

File: packages/ckeditor5-utils/src/madewith.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module utils/madewith
+ */
+
+/**
+ * Possible states of the key after verification
```
Comment (Reinmar) [inline packages/ckeditor5-utils/src/madewith.ts:11]:
```suggestion
 * Possible states of the key after verification.
```

File: packages/ckeditor5-utils/src/madewith.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ */
+export type VerifiedKeyStatus = 'VALID' | 'INVALID' | 'EXPIRED';
```
Comment (Reinmar) [inline packages/ckeditor5-utils/src/madewith.ts:13]: Why do we need EXPIRED? Do we distinguish such a state in the premium features license check mechanism?

File: packages/ckeditor5-utils/src/madewith.ts (hunk 2 above)
Comment (pszczesniak) [reply packages/ckeditor5-utils/src/madewith.ts:22]: I just wanted to add initial version of "friendly comment" ( https://github.com/cksource/ckeditor5-internal/issues/3164#issuecomment-1516053695 ).

File: packages/ckeditor5-utils/src/madewith.ts (hunk 4 above)
Comment (pszczesniak) [reply packages/ckeditor5-utils/src/madewith.ts:13]: I think that when we will have such kind of information we can prepare special url for, for example, expired state - to track it and/or to redirect to some specific landing page or something similar. @mlewand WDYT?

File: packages/ckeditor5-utils/src/madewith.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	if ( decryptedSecondElement.length !== 0x8 ) {
+		return 'INVALID';
+	}
+
+	const day = decryptedSecondElement.substring( 0, 2 );
+	const month = decryptedSecondElement.substring( 2, 4 );
+	const year = decryptedSecondElement.substring( 4, 8 );
+	const date = new Date( `${ year }-${ month }-${ day }` );
```
Comment (pomek) [inline packages/ckeditor5-utils/src/madewith.ts:66]: * Date format: [yyyyMMdd](https://date-fns.org/v2.30.0/docs/format).
* Make sure to decrease month by one when passing it to the `Date` constructor.

Comment (pszczesniak) [reply packages/ckeditor5-utils/src/madewith.ts:66]: > *   Make sure to decrease month by one when passing it to the `Date` constructor.

Nice catch :+1:

Comment (pszczesniak) [reply packages/ckeditor5-utils/src/madewith.ts]: Fixed in e9f4eb0

File: packages/ckeditor5-utils/src/madewith.ts (hunk 4 above)
Comment (pomek) [reply packages/ckeditor5-utils/src/madewith.ts:13]: Available suffixes for premium features: `Valid`, `Invalid`, `Trial:*`

File: packages/ckeditor5-utils/src/madewith.ts (hunk 5 above)
Comment (pszczesniak) [reply packages/ckeditor5-utils/src/madewith.ts]: Date substrings reorder :point\_right: df7f85e

**Comment by mlewand** [comment]
Couple of changes applied as a part of R.

Most importantly improved handling for new format (more checks). Changed fixtures used in unit tests.

**Review by mlewand — APPROVED** [review]

**Pull request merged**